src/internal/storage/fileset/buffer.go

```diff
@@ -25,52 +33,75 @@ func NewBuffer() *Buffer {
 	}
 }
 
-func (b *Buffer) Add(path, datum string) io.Writer {
+func (b *Buffer) addInternal(path, datum string) *file {
 	path = Clean(path, false)
 	if _, ok := b.additive[path]; !ok {
 		b.additive[path] = make(map[string]*file)
 	}
 	datumFiles := b.additive[path]
 	if _, ok := datumFiles[datum]; !ok {
+		b.fileCount++
 		datumFiles[datum] = &file{
 			path:  path,
 			datum: datum,
-			buf:   &bytes.Buffer{},
 		}
 	}
-	f := datumFiles[datum]
-	return f.buf
+	return datumFiles[datum]
+}
+
+func (b *Buffer) Add(path, datum string) io.Writer {
+	f := b.addInternal(path, datum)
+	if len(f.contents) > 0 && f.contents[len(f.contents)-1].copy == nil {
+		return f.contents[len(f.contents)-1].buf
+	}
+	buf := &bytes.Buffer{}
+	f.contents = append(f.contents, fileContent{buf: buf})
+	return buf
 }
 
 func (b *Buffer) Delete(path, datum string) {
 	path = Clean(path, IsDir(path))
 	if IsDir(path) {
 		// TODO: Linear scan for directory delete is less than ideal.
 		// Fine for now since this should be rare and is an in-memory operation.
-		for file := range b.additive {
+		for file, datumFiles := range b.additive {
 			if strings.HasPrefix(file, path) {
+				b.fileCount -= len(datumFiles)
 				delete(b.additive, file)
 			}
 		}
 		return
 	}
 	if datumFiles, ok := b.additive[path]; ok {
+		b.fileCount -= len(datumFiles)
```

Comment:
I don't think this is the correct update. We should decrement by 1 if the path & datum pair existed in the additive set.

Reply:
ah of course, you're right